### DISCUSSION TITLE
The root_path is missing from the redirects from an url with trailing slash to the same url without it

### DISCUSSION THREAD

**Discussion by Mattiam2** [encode/starlette#2514] [Potential Issue]
Issue opened also in the FastAPI discussions [here](https://github.com/tiangolo/fastapi/discussions/11153#discussion-6225371).

### Example Code

```python
from fastapi import FastAPI

app = FastAPI(root_path="/wms")


@app.get("/hello")
async def root():
    return {"message": "Hello World"}
```

The program runs in gunicorn (using the uvicorn worker class) behind a nginx proxy which strips away the root_path (as described in [Behind a proxy](https://fastapi.tiangolo.com/advanced/behind-a-proxy/#setting-the-root_path-in-the-fastapi-app)).

Before fastapi 0.109 (and before starlette 0.35.0) an URL with a trailing slash (ex. https://127.0.0.1/wms/hello/) would be redirected to the corresponding URL without the trailing slash (ex. https://127.0.0.1/wms/hello) as expected.

From fastapi 0.109 (and from starlette 0.35.0) the redirect no more includes the root_path, so from https://127.0.0.1/wms/hello/ the application redirects to https://127.0.0.1/hello which results in a 404 Not Found error.

Here the headers of the redirect request:
![image](https://github.com/encode/starlette/assets/1143780/27344c7d-e875-43ea-9d1b-e1d59cfaa5ed)

I think the issue came from [this change](https://github.com/encode/starlette/commit/c3c6314759b8356df185857c6001de6c8575708d#diff-e939ae940ce35a0db1544bfea4ab853be43ba89ea33e3cac713aa0b19f4bb875) which strips away the root_path from the path in the URL constructor which seems to me off, even considering the ASGI specs.

**Comment by dm1tryG** [comment]
Hello @Mattiam2 I have faced same issue today... Did you know how to pass it?

**Reply by Mattiam2** [reply]
In the meanwhile because I use a reverse proxy (nginx), I've managed to "fix" the issue by specifying the root_path when using proxy_pass with the socket.

Before:
```
location ~ ^/wms/(.*)$ {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    proxy_redirect off;

    if (!-f $request_filename) {
        proxy_pass http://unix:/run/gunicorn/wms.sock:/$1$is_args$args;
        break;
    }
}
```

After:
```
location ~ ^/wms/(.*)$ {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    proxy_redirect off;

    if (!-f $request_filename) {
        proxy_pass http://unix:/run/gunicorn/wms.sock:/wms/$1$is_args$args;
        break;
    }
}
```

Or by just removing the arguments all together (in this way the full path will be used):
`proxy_pass http://unix:/run/gunicorn/wms.sock;`

This solution works enough for me but shouldn't be correct, because for the way root_path is defined within fastapi/starlette, fastapi socket shouldn't expect to have the root_path included in the url path.

**Comment by ScoreX** [comment]
Just chiming in to say this issue is still present.

@Kludex - Should this be raised as an issue to be fixed? if the behaviour is intended, can you recommend a way of making redirects work when using `root_path`?

**Reply by smac89** [reply]
From the looks of things, could very well be a regression, which means it is an issue.
